src/sdk/module/default/simple_storage_bucket.cc

```diff
@@ -149,16 +149,25 @@ void SimpleStorageBucket::read(const string_& key, ReadCallback&& callback,
           }
         }
       } else {
-        LOG_FINE(__FILENAME__,
-                 LOG << "there is no file to read " << key);
+        LOG_FINE(__FILENAME__, LOG << "there is no file to read " << key);
         status = BucketReadStatus::NOT_EXIST;
       }
     } catch (const fs::filesystem_error& ex) {
       LOG_ERROR(__FILENAME__, LOG << "Read failed for " << key << " file");
       status = BucketReadStatus::READ_FAILED;
     }
 
-    callback(std::move(url_decode(key)), std::move(vec), status);
+    boost::asio::io_service::strand* owner_strand = get_owner_strand();
+
+    if (owner_strand != nullptr)
+      owner_strand->post([
+        callback = std::move(callback), decode_key = std::move(url_decode(key)),
+        vec = std::move(vec), status = std::move(status)
+      ]() {
+        callback(std::move(decode_key), std::move(vec), std::move(status));
+      });
+    else
+      callback(std::move(url_decode(key)), std::move(vec), std::move(status));
   };
```

Comment:
we need a wrapper class to handle it in storage.cc, otherwise every storage implementation like this will need to have the same code to if else on strand post